packages/react-utilities/src/descendants/descendants.ts

```diff
@@ -0,0 +1,72 @@
+import * as React from 'react';
+import { useId, useFirstMount, useIsomorphicLayoutEffect } from '../hooks/index';
+
+export type Descendant = { id: string; forceUpdate: () => void };
+export type Descendants = Record<string, Descendant>;
+export type SetDescendant = (descendant: Descendant) => number;
+
+export interface DescendantsContextValue {
+  /**
+   * A record of descendants that have been registered by their id
+   */
+  descendants: Descendants;
+  /**
+   * add a descendant, if it hasn't been added before
+   * @returns index of descendant
+   */
+  setDescendant: SetDescendant;
+}
+
+const DescendantsContext = React.createContext<DescendantsContextValue | undefined>(undefined);
+
+/**
+ * hook to initialize descendants,
+ * this hook ensures that although new descendants can be added,
+ * no re-rendering will be triggered.
+ * The only way to trigger a re-rendering is by updating the parent itself,
+ * e.g: adding a new descendant, removing a descendant, re-ordering descendants, ...
+ * @returns the record of descendants and a function to add new descendants
+ */
+export const useDescendants = () => {
+  const isFirstMount = useFirstMount();
+  const descendants = React.useRef<Descendants>({});
+  const order = React.useRef<string[]>([]);
```

Comment:
If you use a `Map` you might not need to use `order` since it remembers the insertion order of keys: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

Reply:
The problem with using a `Map` in this scenario is the necessity for creating an `Array` out of the map every time an insertion is done. I'm not sure if it would be worth it.